### PR TITLE
refactor(agent): use semantic drafts in structural skills

### PR DESCRIPTION
**Pull request by qinsz01**
## Summary

- Problem: structural skills still depended on brittle natural-language parser fallbacks, and benchmark runs exposed skill coverage gaps across beam, frame, truss, concrete-frame, and column cases.
- Why it matters: these parser paths made results harder to attribute to model capability and could drop structured draft issues or wind parameters during retries / multi-turn updates.
- What changed: added semantic engineering draft normalization, routed core structural skills through structured draft output, removed major natural-parser fallbacks, improved model-builder coverage, added column skill support, preserved draft issues, and preserved frame wind state.
- What did NOT change (scope boundary): no frontend changes, no user settings/config changes, no benchmark submodule pointer update, no secrets or API keys.

## Change Type (select all)

- [x] Bug fix
- [ ] Feature
- [x] Refactor required for the fix
- [ ] Docs
- [ ] Security hardening
- [ ] Chore / infra

## Scope (select all touched areas)

- [x] Gateway / orchestration
- [x] Skills / tool execution
- [ ] Auth / tokens
- [ ] Memory / storage
- [ ] Integrations
- [ ] API / contracts
- [ ] UI / DX
- [ ] CI/CD / infra

## Linked Issue/PR

- Closes #
- Related #
- [x] This PR fixes a bug or regression

## Root Cause / Regression History (if applicable)

- Root cause: core structural skills mixed semantic LLM draft extraction with local natural-language parser fallbacks; several runtime paths did not consistently preserve structured draft issues or frame wind parameters.
- Missing detection / guardrail: existing unit coverage did not lock down semantic draft consumption, invalid-value issue preservation, or wind state merging.
- Prior context: found while calibrating StructureClaw benchmark standard and interaction suites.
- Why this regressed now: the benchmark started exercising more skill-specific cases and retry/interaction paths.
- If unknown, what was ruled out: benchmark-only evaluator issues were separated from main repository runtime/skill behavior before this fix.

## Regression Test Plan (if applicable)

- Coverage level that should have caught this:
  - [x] Unit test
  - [x] Seam / integration test
  - [ ] End-to-end test
  - [ ] Existing coverage already sufficient
- Target test or file:
  - `backend/src/agent-langgraph/__tests__/param-extractor.test.mjs`
  - `backend/src/agent-runtime/__tests__/runtime-helpers.test.mjs`
  - `backend/src/agent-skills/structure-type/*/__tests__/handler.test.mjs`
- Scenario the test should lock in: semantic draft extraction, draft issue preservation, frame wind state merge, and skill model coverage for beam/frame/concrete-frame/truss/column cases.
- Why this is the smallest reliable guardrail: it validates the skill/runtime boundary without requiring live LLM calls.
- Existing test that already covers this (if any): N/A.
- If no new test is added, why not: N/A.

## User-visible / Behavior Changes

Structural skill extraction now relies more on structured semantic draft output instead of brittle local natural-language parser fallbacks. Invalid structured values can surface as draft issues instead of being silently normalized away. Frame wind pressure can be retained through draft merge and converted into lateral floor loads for supported frame cases.

## Diagram (if applicable)

```text
Before:
[user message] -> [LLM draft] -> [natural parser fallback] -> [skill model]

After:
[user message] -> [semantic engineering draft] -> [skill normalization] -> [skill model]
```

## Security Impact (required)

- New permissions/capabilities? (`Yes/No`) No
- Secrets/tokens handling changed? (`Yes/No`) No
- New/changed network calls? (`Yes/No`) No
- Command/tool execution surface changed? (`Yes/No`) No
- Data access scope changed? (`Yes/No`) No
- If any `Yes`, explain risk + mitigation: N/A

### Environment

- OS: Windows
- Runtime/container: local Node/npm backend workspace
- Model/provider: N/A for unit validation
- Integration channel (if any): N/A
- Relevant config (redacted): N/A

### Steps

1. Build backend TypeScript.
2. Run affected agent runtime and structural skill unit tests.
3. Run agent orchestration contract validation.

### Expected

- Backend compiles.
- Affected tests pass.
- Agent orchestration validation passes.

### Actual

- Backend compiled successfully.
- 7 targeted Jest suites passed, 53 tests total.
- Agent orchestration validation passed.

### Evidence

- [x] Failing test/log before + passing after
- [x] Trace/log snippets
- [ ] Screenshot/recording
- [ ] Perf numbers (if relevant)

Commands run:

```text
npm run build --prefix backend
npm test --prefix backend -- --runInBand src/agent-langgraph/__tests__/param-extractor.test.mjs src/agent-runtime/__tests__/runtime-helpers.test.mjs src/agent-skills/structure-type/beam/__tests__/handler.test.mjs src/agent-skills/structure-type/frame/__tests__/handler.test.mjs src/agent-skills/structure-type/concrete-frame/__tests__/handler.test.mjs src/agent-skills/structure-type/truss/__tests__/handler.test.mjs src/agent-skills/structure-type/column/__tests__/handler.test.mjs
node tests/runner.mjs validate validate-agent-orchestration
```

## Human Verification (required)

What you personally verified (not just CI), and how:

- Verified scenarios: semantic draft extraction paths, affected structure skill handler tests, frame wind state preservation, and agent orchestration metadata/contracts.
- Edge cases checked: invalid draft issue preservation, multi-story frame wind merge, truss/column skill coverage, concrete-frame semantic draft paths.
- What you did **not** verify: full live LLM benchmark rerun in this PR validation pass.

## Review Conversations

- [ ] I replied to or resolved every bot review conversation I addressed in this PR.
- [ ] I left unresolved only the conversations that still need reviewer or maintainer judgment.

N/A for initial PR creation.

## Compatibility / Migration

- Backward compatible? (`Yes/No`) Yes
- Config/env changes? (`Yes/No`) No
- Migration needed? (`Yes/No`) No
- If yes, exact upgrade steps: N/A

## Risks and Mitigations

- Risk: semantic draft quality depends on model output quality, so some benchmark failures may now reflect model capability more directly.
  - Mitigation: added unit coverage for normalization and guardrails; benchmark should record failures rather than requiring all cases to pass.
- Risk: wind-to-floor-load conversion is currently a simple supported-frame approximation.
  - Mitigation: retained structured wind state so future PKPM/YJK or richer frame logic can consume the original parameter.

---

> [!NOTE]
> **Medium Risk**
> Large refactor of draft merge, validation, and model-building paths for core structural agent flows; behavior shifts from parser fallbacks to LLM-structured drafts, with broad unit test coverage but high regression surface on multi-turn and frame load logic.
> 
> **Overview**
> Structural parameter extraction and skills now center on **`engineeringDraft`** (geometry, loads, wind, analysis) with optional legacy **`draftPatch`**, plus **`draftIssues`** and **`skillState.invalidDraftFields`** for bad or ambiguous values instead of silently normalizing them.
> 
> A new **`engineering-draft`** runtime layer normalizes and merges semantic drafts, projects frame line/area/wind loads into per-story **`floorLoads`**, and feeds expanded **model builders** (semantic beams, columns, continuous beams, multi-bay portal frames). **Concrete-frame** paths drop major **natural-language parser** fallbacks in favor of structured LLM output; **beam** and new **column** skills route through projection instead of message-based defaults. **Type detection** keeps the current structural type on parameter-update turns when draft state is stable.
> 
> LangGraph **param-extractor** / **executor** prompts and JSON unwrapping accept top-level semantic JSON; **`extract_draft_params`** and **`detectStructuralType`** pass draft context and expose extraction metadata.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit eac6fe839cbe77f6c1153dda5836dedc83ec6a34. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: backend/src/agent-skills/structure-type/column/handler.ts [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+  }
+  if (nextPatch.loadKN !== undefined) {
+    nextPatch.loadType = nextPatch.loadType ?? 'point';
+    nextPatch.loadPosition = nextPatch.loadPosition ?? 'top-nodes';
+  }
+  if (semanticPatch.skillState) {
+    nextPatch.skillState = semanticPatch.skillState;
+  }
+  return nextPatch;
+}
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-skills/structure-type/column/handler.ts:48]: 合理，已在 `eac6fe8` 中保留 `semanticPatch.draftIssues`，并补了 column handler 单测覆盖。

File: backend/src/agent-runtime/registry.ts [hunk 2]
```diff
@@ -5,6 +5,35 @@ import { buildUnknownStructuralType, detectUnsupportedStructuralTypeByRules } fr
...
 
+function hasStableCurrentState(state: DraftState | undefined): state is DraftState {
+  return !!state?.inferredType && state.inferredType !== 'unknown';
+}
+
+function isExplicitStructuralSwitch(message: string): boolean {
+  const text = message.toLowerCase();
+  return /(?:改为|改成|切换为|换成|按|作为|用).*(?:梁|桁架|门式刚架|门架|钢框架|混凝土框架|框架|柱)/u.test(message)
+    || /(?:change|switch|convert|use|as).*(?:to|into|as).*(?:beam|truss|portal|frame|column)/i.test(text);
+}
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-runtime/registry.ts:16]: 合理，已在 `eac6fe8` 中收紧 structural switch 正则，避免把普通参数更新误判成结构类型切换，并补了 `change height to 4m for the column` 的回归测试。

File: backend/src/agent-runtime/fallback.ts [hunk 3]
```diff
@@ -78,6 +80,63 @@ function mergeFloorLoads(
...
+    }
+  }
+  for (const field of ['storyHeightsM', 'bayWidthsM', 'bayWidthsXM', 'bayWidthsYM'] as const) {
+    const value = patch[field];
+    if (Array.isArray(value) && value.length > 0 && value.every((item) => typeof item === 'number' && Number.isFinite(item) && item > 0)) {
+      valid.add(field);
+    }
+  }
+  return valid;
+}
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-runtime/fallback.ts:123]: 合理，已在 `eac6fe8` 中让 `collectValidPatchFields` 读取 `engineeringDraft.geometry.spanLengthsM` 等数组字段，合法补参后可以清理旧 invalid issue，并加了回归测试。

File: backend/src/agent-runtime/engineering-draft.ts [hunk 4]
```diff
@@ -0,0 +1,713 @@
...
+        stories = [sameFieldUntargetedLoads.findIndex((item) => item.index === index) + 1];
+      } else {
+        stories = Array.from({ length: storyCount }, (_, storyIndex) => storyIndex + 1);
+      }
+    }
+
+    for (const story of stories) {
+      assignFloorLoadValue(floorLoadsByStory, story, field, Number(totalKN.toFixed(6)));
+    }
+  }
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-runtime/engineering-draft.ts:565]: 合理，已在 `eac6fe8` 中修复多个未指定楼层荷载少于层数时被复制到所有楼层的问题。实现上保留单个未指定楼层荷载作用于全部楼层的旧语义，多个未指定荷载则按顺序映射，并加了回归测试。

File: backend/src/agent-runtime/engineering-draft.ts [hunk 5]
```diff
@@ -0,0 +1,713 @@
...
+    五: 5,
+    六: 6,
+    七: 7,
+    八: 8,
+    九: 9,
+    十: 10,
+  };
+  const parsed = chineseMatch?.[1] ? table[chineseMatch[1]] : undefined;
+  return parsed !== undefined && parsed <= storyCount ? parsed : undefined;
+}
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-runtime/engineering-draft.ts:435]: 合理，已在 `eac6fe8` 中支持复合中文楼层序号，包括 `十一`、`两`、`廿` 等，并补了 `第十一层` 定向荷载测试。

File: backend/src/agent-runtime/model-builder.ts [hunk 6]
```diff
@@ -458,6 +478,451 @@ function buildTrussModel(state: DraftState, metadata: Record<string, unknown>):
...
+  const mezzanineLoad = readPositiveNumber(state.skillState?.mezzanineLoadKN);
+  const hasMezzanine = mezzanineHeight !== undefined && mezzanineHeight > 0 && mezzanineHeight < height && spans.length >= 1;
+  const nodes: Array<Record<string, unknown>> = [];
+  const elements: Array<Record<string, unknown>> = [];
+  const loads: Array<Record<string, unknown>> = [];
+
+  for (let i = 0; i < xCoordinates.length; i += 1) {
+    nodes.push({ id: `B${i}`, x: xCoordinates[i], y: 0, z: 0, restraints: [true, true, true, true, true, true] });
+    nodes.push({ id: `T${i}`, x: xCoordinates[i], y: 0, z: height });
+  }
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-runtime/model-builder.ts:861]: 合理，已在 `eac6fe8` 中让 portal-frame model builder 使用 `frameBaseSupportType`。pinned 的 DOF 取值按现有 frame / concrete-frame base restraint 约定保持一致，并加了 pinned base 单测。

File: backend/src/agent-skills/structure-type/beam/handler.ts [hunk 7]
```diff
@@ -45,13 +39,21 @@ function applyBeamDefaults(message: string, patch: DraftExtraction): DraftExtrac
...
   });
-  return restrictLegacyDraftPatch(applyBeamDefaults(message, domainPatch), 'beam', [...ALLOWED_KEYS]);
+  const nextPatch = restrictLegacyDraftPatch(applyBeamDefaults(domainPatch), 'beam', [...ALLOWED_KEYS]);
+  if (semanticPatch.engineeringDraft) {
+    nextPatch.engineeringDraft = semanticPatch.engineeringDraft;
+  }
+  if (semanticPatch.skillState) {
+    nextPatch.skillState = semanticPatch.skillState;
+  }
+  return nextPatch;
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-skills/structure-type/beam/handler.ts:56]: 合理。已在 `efbf3bf` 中让 `composeStructuralDomainPatch` 保留 `draftIssues`/`skillState`/`engineeringDraft`，覆盖 beam 以及同类结构 handler，并补了 beam 的 draftIssues 保留测试。

File: backend/src/agent-runtime/engineering-draft.ts [hunk 8]
```diff
@@ -0,0 +1,731 @@
...
+  return {
+    structureType: patch.structureType ?? existing.structureType,
+    geometry: { ...(existing.geometry ?? {}), ...(patch.geometry ?? {}) },
+    material: { ...(existing.material ?? {}), ...(patch.material ?? {}) },
+    sections: { ...(existing.sections ?? {}), ...(patch.sections ?? {}) },
+    boundary: { ...(existing.boundary ?? {}), ...(patch.boundary ?? {}) },
+    loads: patch.loads?.length ? patch.loads : existing.loads,
+    wind,
+    analysis: { ...(existing.analysis ?? {}), ...(patch.analysis ?? {}) },
+  };
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-runtime/engineering-draft.ts:278]: 合理。已在 `efbf3bf` 中改为按 load 的 kind/unit/direction/target/location 语义键合并 engineeringDraft loads；同一语义位置的新 load 会替换旧值，不同 load 会保留，避免多轮更新丢荷载或重复荷载。

File: backend/src/agent-runtime/engineering-draft.ts [hunk 9]
```diff
@@ -0,0 +1,731 @@
...
+      if (sameFieldUntargetedLoads.length > 1 && sameFieldUntargetedLoads.length <= storyCount) {
+        stories = untargetedIndex >= 0 ? [untargetedIndex + 1] : [];
+      } else {
+        stories = Array.from({ length: storyCount }, (_, storyIndex) => storyIndex + 1);
+      }
+    }
+
+    for (const story of stories) {
+      assignFloorLoadValue(floorLoadsByStory, story, field, Number(totalKN.toFixed(6)));
+    }
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-runtime/engineering-draft.ts:564]: 合理。已在 `efbf3bf` 中修复超过楼层数的未指定楼层荷载：多个未指定荷载按顺序映射到可用楼层，超出的不再复制到所有楼层，并加了回归测试。

File: backend/src/agent-runtime/fallback.ts [hunk 10]
```diff
@@ -343,24 +413,28 @@ export function mergeDraftState(existing: DraftState | undefined, patch: DraftEx
...
+    ? {
+      ...skillStateWithoutInvalid,
       ...(invalidDraftFields ? { invalidDraftFields } : {}),
     }
     : undefined;
 
   return {
     inferredType: mergedType,
+    engineeringDraft,
+    draftIssues,
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-runtime/fallback.ts:437]: 合理。已在 `efbf3bf` 中避免把 frame 的 line/area load 原始强度写成 legacy `loadKN`，这样在 floorLoads 投影失败时不会把 kN/m 或 kN/m² 误当成每层总 kN 回填，并加了测试。

File: backend/src/agent-runtime/registry.ts [hunk 11]
```diff
@@ -5,6 +5,35 @@ import { buildUnknownStructuralType, detectUnsupportedStructuralTypeByRules } fr
...
+function isExplicitStructuralSwitch(message: string): boolean {
+  const text = message.toLowerCase();
+  return /(?:改为|改成|切换为|换成|按|作为|用)\s*(?:梁|桁架|门式刚架|门架|钢框架|混凝土框架|框架|柱)/u.test(message)
+    || /(?:change|switch|convert)\s+(?:to|into|as)\s+(?:beam|truss|portal|frame|column)\b/i.test(text);
+}
+
+function looksLikeCurrentDraftUpdate(message: string): boolean {
+  const text = message.toLowerCase();
+  return /(?:刚才|当前|原来|继续|再|另外|同时|补充|考虑|增加|添加|调整|修改|改成|改为|改到|变成|换成|设为)/u.test(message)
+    || /\b(?:previous|current|same|continue|also|add|include|update|modify|change|adjust|set)\b/i.test(text);
```
[the remark this answers is not in the document]
Comment (qinsz01) [reply backend/src/agent-runtime/registry.ts:21]: 合理。已在 `efbf3bf` 中支持 `change to a beam` 这类带冠词的英文显式切换，并在显式切换时不再把 currentState 传给各 skill detector，避免旧 frame fallback 抢回路由；已加回归测试。

**Review by guyi2000 — APPROVED** [review]
